core/src/main/java/org/apache/iceberg/io/SortedPosDeleteWriter.java

```diff
@@ -0,0 +1,172 @@
+                        long recordsNumThreshold) {
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.format = format;
+    this.partition = partition;
+    this.recordsNumThreshold = recordsNumThreshold;
+  }
+
+  SortedPosDeleteWriter(FileAppenderFactory<T> appenderFactory,
+                        OutputFileFactory fileFactory,
+                        FileFormat format,
+                        PartitionKey partition) {
+    this(appenderFactory, fileFactory, format, partition, DEFAULT_RECORDS_NUM_THRESHOLD);
+  }
+
+  public void delete(CharSequence path, long pos) {
+    delete(path, pos, null);
+  }
+
+  public void delete(CharSequence path, long pos, T row) {
+    posDeletes.put(CharSequenceWrapper.wrap(path), PosValue.of(pos, row));
+    records += 1;
+
+    // TODO Flush buffer based on the policy that checking whether whole heap memory size exceed the threshold.
+    if (records >= recordsNumThreshold) {
+      flushBuffer();
+    }
+  }
+
+  public List<DeleteFile> complete() throws IOException {
+    close();
+
+    return completedFiles;
+  }
+
+  public Set<CharSequence> referencedDataFiles() {
+    return referencedDataFiles;
+  }
+
+  @Override
+  public void close() throws IOException {
+    flushBuffer();
+  }
+
+  private void flushBuffer() {
+    if (posDeletes.isEmpty()) {
+      return;
+    }
+
+    // Create a new output file.
+    EncryptedOutputFile outputFile;
+    if (partition == null) {
+      outputFile = fileFactory.newOutputFile();
+    } else {
+      outputFile = fileFactory.newOutputFile(partition);
+    }
+
+    PositionDeleteWriter<T> writer = appenderFactory.newPosDeleteWriter(outputFile, format, partition);
+    try (PositionDeleteWriter<T> closeableWriter = writer) {
+      // Sort all the paths.
+      List<CharSequence> paths = Lists.newArrayListWithCapacity(posDeletes.keySet().size());
+      for (CharSequenceWrapper charSequenceWrapper : posDeletes.keySet()) {
+        paths.add(charSequenceWrapper.get());
+      }
+      paths.sort(Comparators.charSequences());
+
+      // Write all the sorted <path, pos, row> triples.
+      for (CharSequence path : paths) {
+        List<PosValue<T>> positions = (List<PosValue<T>>) posDeletes.get(wrapper.set(path));
+        positions.sort(Comparator.comparingLong(PosValue::pos));
+
+        positions.forEach(posValue -> closeableWriter.delete(path, posValue.pos(), posValue.row()));
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to write the sorted path/pos pairs to pos-delete file: " +
+          outputFile.encryptingOutputFile().location(), e);
+    }
+
+    // Clear the buffered pos-deletions.
+    posDeletes.clear();
+    records = 0;
+
+    // Add the referenced data files.
+    referencedDataFiles.addAll(writer.referencedDataFiles());
+
+    // Add the completed delete files.
+    completedFiles.add(writer.toDeleteFile());
+  }
+
+  private static class PosValue<R> {
```

Comment:
Minor: This is called `PosValue`, but the values is always used as "row" so it would make more sense to call it `PosRow` or `PosAndRow`. That way references use `posRow.row()` rather than `posValue.row()`.